spec/abstract-jobs.html

```diff
@@ -275,33 +275,17 @@ <h1>AddToKeptObjects ( _object_ )</h1>
     </emu-note>
   </emu-clause>
 
-  <emu-clause id="sec-check-for-empty-cells" aoid="CheckForEmptyCells">
-    <h1> CheckForEmptyCells ( _finalizationRegistry_ ) </h1>
-    <p> The following steps are performed: </p>
-    <emu-alg>
-      1. Assert: _finalizationRegistry_ has an [[Cells]] internal slot.
-      1. For each _cell_ in _finalizationRegistry_.[[Cells]], do
-        1. If _cell_.[[WeakRefTarget]] is ~empty~, then
-          1. Return *true*.
-      1. Return *false*.
-    </emu-alg>
-  </emu-clause>
-
   <emu-clause id="sec-cleanup-finalization-registry" aoid="CleanupFinalizationRegistry">
     <h1> CleanupFinalizationRegistry ( _finalizationRegistry_ [ , _callback_ ] ) </h1>
     <p> The following steps are performed: </p>
     <emu-alg>
-      1. Assert: _finalizationRegistry_ has [[Cells]],
-      [[CleanupCallback]], and [[IsFinalizationRegistryCleanupJobActive]] internal slots.
-      1. If CheckForEmptyCells(_finalizationRegistry_) is *false*, return.
-      1. Let _iterator_ be ! CreateFinalizationRegistryCleanupIterator(_finalizationRegistry_).
+      1. Assert: _finalizationRegistry_ has [[Cells]] and [[CleanupCallback]] internal slots.
       1. If _callback_ is not present or *undefined*, set _callback_ to _finalizationRegistry_.[[CleanupCallback]].
-      1. Set _finalizationRegistry_.[[IsFinalizationRegistryCleanupJobActive]] to *true*.
-      1. Let _result_ be Call(_callback_, *undefined*, &laquo; _iterator_ &raquo;).
-      1. Set _finalizationRegistry_.[[IsFinalizationRegistryCleanupJobActive]] to *false*.
-      1. Set _iterator_.[[FinalizationRegistry]] to ~empty~.
-      1. If _result_ is an abrupt completion, return _result_.
-      1. Else, return NormalCompletion(*undefined*).
+      1. While _finalizationRegistry_.[[Cells]] contains a Record _cell_ such that _cell_.[[WeakRefTarget]] is ~empty~, then an implementation may perform the following steps,
+        1. Choose any such _cell_.
+        1. Remove _cell_ from _finalizationRegistry_.[[Cells]].
+        1. Perform ? Call(_callback_, *undefined*, &laquo; _cell_.[[HeldValue]] &raquo;).
```

Comment:
Reading this again, this termination behavior isn't quite what I envisioned wrt backpressure on the cleanup tasks.

It's correct for cleanupSome to propagate the error up, but for the cleanup task, @mhofman's right in that letting exceptions interrupt is technically the same as letting the user express backpressure to the engine. The question to the engine is the same: if it has an interrupted iteration in the cleanup task, what should it do with that FR? Should the engine reschedule it? What if the user cleanup function always throws? Should the engine drop the FR after N tries?

I was envisioning something like in the background task, that all [[Cells]] in the FR will be iterated even if individual calls to cleanup throw. `window.onerror` will be called once each for the error.

Reply:
Actually, I guess according to the current spec whether an implementation interrupts iteration or not is indistinguishable from an implementation that doesn't. The implementation that doesn't interrupt is equivalent to immediately running cleanup again after the current iteration is interrupted.

If this is intentional (which I'm happy with if so), then this deserves a note.

Reply:
Ah wait, no, it's distinguishable in observing when the microtask checkpoint is performed...

Reply:
I think cleanupSome will just trigger one iteration, so it's even more distinguishable. With the automatically scheduled thing, yeah, we don't give flexibility on the microtask checkpoint around reporting an error that we could decide to give.

One path here would be to collect all the errors and throw an AggregateError if there's more than one (regardless of whether it's from cleanupSome or the background scheduler). What do you think of those semantics?

Reply:
After thinking about this more and discussing it with @syg offline, my (our?) current thought is:
- AggregateError is unnecessary complexity; just throwing the first exception as the current PR does is a good solution.
- It's OK to exit iteration early due to an exception, as long as we keep making forward progress. Each time, we'll have at least one iteration, so that's guaranteed (both with cleanupSome and with background calls).
- Implementations could maintain the logic for background threads, to keep calling the cleanup callback with all the items they have, with the simple addition of a microtask checkpoint required after any exception thrown (which they are allowed to do anyway even if there's no exception). With this strategy, no requeueing is required.